src/radical/pilot/utils/prof_utils.py

```diff
@@ -1236,5 +1236,160 @@ def _get_unit_consumption(session, unit, udurations=None):
     return ret
 
 
+# ------------------------------------------------------------------------------
+#
+def get_resource_transitions(pilot, task_metrics=None, pilot_metrics=None):
+
+    if not task_metrics:
+        if pilot.cfg['task_launch_method'] == 'PRTE':
+            task_metrics = UNIT_DURATIONS_PRTE
+        else:
+            task_metrics = UNIT_DURATIONS_DEFAULT
+
+    if not pilot_metrics:
+        pilot_metrics = PILOT_DURATIONS
+
+
+    # we try to find points in time where resource usage moved from purpose A to
+    # purpose B.  For example, consider this metric:
+    #
+    #   'exec_queue'  : [{ru.EVENT: 'schedule_ok'            },
+    #                    {ru.STATE: s.AGENT_EXECUTING        }],
+    #   'exec_prep'   : [{ru.STATE: s.AGENT_EXECUTING        },
+    #                    {ru.EVENT: 'exec_start'             }],
+    #   'exec_rp'     : [{ru.EVENT: 'exec_start'             },
+    #                    {ru.EVENT: 'cu_start'               }],
+    #
+    # then we convert this into the following structure:
+    #
+    # [
+    #   # event                          from         to
+    #   [{ru.EVENT: 'schedule_ok'    },  None,        'exec_queue'],
+    #   [{ru.STATE: s.AGENT_EXECUTING}, 'exec_queue', 'exec_prep'],
+    #   [{ru.EVENT: 'exec_start'     }, 'exec_prep',  'exec_rp'],
+    #   [{ru.EVENT: 'cu_start'       }, 'exec_rp',    'None]
+    # ]
+    #
+    # which will use like this:
+    #
+    #   - go through all events for a task
+    #   - if event is in list above
+    #     - reduce resources in `from` metric
+    #     - increase resource in `to` metric
+    #
+    # If `from` or `to` are None, then the resources are taken from / given to
+    # the pilot's idle resources.  We thus rename the 'None' to 'idle'.
+
```

Comment:
Does that assume a 'perfect' profile in which events are not missing? That is a fairly rare event.

Reply:
No, that mechanism also works for missing events.  I mean, we then obviously miss the respective resource shifts.  For example, if `cu_exec_stop` is missing we would implicitly delay the change of resource ownership from the executor to the scheduler - but that's to be expected, right?

Reply:
Yes, it is expected but it will drug the profile and might be difficult to understand that it was indeed drugged. I guess we will have to catch that in analytics.

Reply:
Yes - completeness of events has to be caught elsewhere - the plotting can't do anything about missing information...